### ISSUE TITLE
Some questions aboout dataset

### ISSUE THREAD

**Issue by championbridge** [JiuhaiChen/BLIP3o#11]
1.Why is preprocessing through gen_vision still required for images in the understanding (und) task? Shouldn't the model only need to process the image for understanding without the need to generate new images?
`elif inst_type == "und":

                    resized_images = [transform_und_images(img) for img in images]

                    image_inputs = self.data_args.image_processor(resized_images, return_tensors="pt")

                    data_dict["und_image"] = image_inputs.pixel_values
                    data_dict["grid_thw"] = image_inputs.image_grid_thw
                    data_dict["gen_image"] = img_process(
                        resized_images,
                        self.data_args.gen_image_processor,
                        self.data_args.image_aspect_ratio,
                    )`
2.Additionally, what is the purpose of the data_dict["image"] field?

3.Why is there no <image> placeholder in the value for the I2I task? Shouldn't the I2I task also include <image> to indicate the image input for the model?
`elif sources["type"] == "I2I" or sources["type"] == "journeyDB_I2I":
    sources["conversations"] = [
        {"from": "human", "value": f"<image>\nPlease reconstruct the given image."},
        {"from": "gpt", "value": ""},
    ]`

**Comment by JiuhaiChen** [comment]
Hi, thanks! We are still trying to clean up this part of data process, sorry for the confusion. 
```
                resized_images = [transform_und_images(img) for img in images]

                image_inputs = self.data_args.image_processor(resized_images, return_tensors="pt")

                data_dict["und_image"] = image_inputs.pixel_values
                data_dict["grid_thw"] = image_inputs.image_grid_thw
                data_dict["gen_image"] = img_process(
                    resized_images,
                    self.data_args.gen_image_processor,
                    self.data_args.image_aspect_ratio,
                )`
```
This part of data process is for the `sources["type"] == "I2I"`,  image is input from understanding vision encoder, so we have data_dict["und_image"] and data_dict["grid_thw"], also the image should be reconstructed from the generation vision decoder, so we need  data_dict["gen_image"]. Note here  data_dict["gen_image"]  and data_dict["und_image"]  are processing the same images. This is not image editing task, just for image reconstruction. 

2. We are checking the code, maybe some redundant variable. 
3. Yeah, you are correct, we should put a place holder in the value.

There are some minor issues for I2I, we are also actively fixing it. But for the T2I, it should be good.

**Comment by JiuhaiChen** [comment]
Thanks again for your interest! The codebase is fairly large, so here’s a high-level overview:

The overall structure is similar to LLaVA’s, if you’ve worked with LLaVA before, you’ll find it familiar! 

1. Data loading & preprocessing (all in train.py):

Image tar loading: https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L572

Conversation formatting: https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L619

Image preprocessing for CLIP based generation: https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L721

Tokenization: https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L715



2. Image-generation pipeline
We format the input as 
```
<text>…</text><[IMG]><img><img>…<img><[/IMG]>
```

```<text>``` contains your prompt (for example, a cute cat) and chat template (for example, Qwen Chat template).

We reserve 64 ```<img>``` placeholder image tokens per image.

<[IMG]> and <[/IMG]> mark the start and end of the image token block.

During training, the whole 64 ```<img>``` is replaced by a learned latent query vector – see
https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/model/blip3o_arch.py#L356

After autoregressive model forwards, you’ll get predicted visual features (the latent queries), and we compute the diffusion loss here:
https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/model/language_model/blip3o_qwen.py#L164

3. Diffusion-transformer architecture
All of diffusion transformers are defined in:
https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/model/nextdit_crossattn.py

Let me know if you’d like more detail on any part!

**Comment by championbridge** [comment]
Thank you for the author's patient reply. So, in the case of I2I, the input image is used as und_image, and the output image (ground truth) should be gen_image. Also, I have a question regarding the dataset. The image column in the traindataset initialized by LazySupervisedMixDataset should be images read using PIL, right?
`train_dataset = load_dataset("webdataset", data_files=data_files, split="train", num_proc=128)
        train_dataset = train_dataset.rename_column("jpg", "image")
        train_dataset = train_dataset.add_column('type', len(train_dataset) * ['T2I'])
        train_dataset = train_dataset.add_column('image_path', len(train_dataset) * [None])
        train_dataset = train_dataset.remove_columns([col for col in train_dataset.column_names if not col in (
            ["image", "txt", "type", "image_path"])])`

**Comment by JiuhaiChen** [comment]
Hi, for the webdataset, you don’t need to read the image, as shown https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L679
You can directly process the image.

**Comment by championbridge** [comment]
If I use JSON to construct the dataset during training instead of using the webdataset format, do the image paths in the JSON need to be read with PIL and then placed into the dataset
Here is the rewritten code.
`image_paths = item["image"]  # List with one path, e.g., ["/path/to/image.png"]
           image_path = image_paths[0]
           img = Image.open(image_path)
           formatted_data.append({
                "image": img,  # Single PIL image object
                "txt": txt,  # Text corresponding to the image
                "type": "T2I",  # Type of the item, e.g., "T2I"
                "image_path": image_path  # Path to the image file
            })`

**Comment by JiuhaiChen** [comment]
hi, this is loading the JourneyDB json file:

```
        train_dataset = load_dataset("json", data_files='/fsx/sfr/data/jiuhai/hub/datasets--JourneyDB--JourneyDB/snapshots/e191aa61ca37e5e4418707ade4df5deb5c6d5d8f/data/train/train_caption_only.jsonl', split="train", num_proc=64)
        train_dataset = train_dataset.add_column('type', len(train_dataset) * ['journeyDB_T2I'])
        train_dataset = train_dataset.add_column('image', len(train_dataset) * [None])
        train_dataset = train_dataset.rename_column("caption", "txt")
        train_dataset = train_dataset.rename_column("img_path", "image_path")
        train_dataset = train_dataset.remove_columns([col for col in train_dataset.column_names if not col in (
            ["txt", "image", "type", "image_path"])])
        print(f"finish loading journeyDB {len(train_dataset)}")
```

Since we need to mix the different data type, webdataset and json, we need to make sure all the column name are the same if we concatenate two different type data. This is reason we put train_dataset = train_dataset.add_column('image', len(train_dataset) * [None]) here,

If you are using json data, you need to read the image with PIL. 
https://github.com/JiuhaiChen/BLIP3o/blob/main/blip3o/train/train.py#L682
